clients/fides-js/src/lib/gpp/string-to-consent.ts

```diff
@@ -0,0 +1,308 @@
+import { updateConsentPreferences } from "../preferences";
+import { EMPTY_ENABLED_IDS } from "../tcf/constants";
+import { EnabledIds, TcfSavePreferences } from "../tcf/types";
+import {
+  createTCFConsentPreferencesToSave,
+  createTcfSavePayload,
+  createTcfSavePayloadFromMinExp,
+  updateTCFCookie,
+} from "../tcf/utils";
+import { FIDES_US_REGION_TO_GPP_SECTION } from "./constants";
+import { GPPUSApproach } from "./types";
+import { deriveGppFieldRegion, US_NATIONAL_REGION } from "./us-notices";
+
+const FORBIDDEN_LEGITIMATE_INTEREST_PURPOSE_IDS = [1, 3, 4, 5, 6];
+
+/**
+ * Gets the consent object from a GPP CMP API model based on the experience configuration
+ *
+ * This is the inverse of setGppOptOutsFromCookieAndExperience
+ */
+const getConsentFromGppCmpApi = ({
+  cmpApi,
+  experience,
+}: {
+  cmpApi: CmpApi;
+  experience: PrivacyExperience;
+}): NoticeConsent => {
+  const consent: NoticeConsent = {};
+  const {
+    privacy_notices: notices = [],
+    region: experienceRegion,
+    gpp_settings: gppSettings,
+  } = experience;
+  const usApproach = gppSettings?.us_approach;
+  let gppRegion = deriveGppFieldRegion({
+    experienceRegion,
+    usApproach,
+  });
+  let gppSection = FIDES_US_REGION_TO_GPP_SECTION[gppRegion];
+
+  if (!gppSection && usApproach === GPPUSApproach.ALL) {
+    // if we're using the all approach, and the current state isn't supported, we should default to national
+    gppRegion = US_NATIONAL_REGION;
+    gppSection = FIDES_US_REGION_TO_GPP_SECTION[gppRegion];
+  }
+
+  if (
+    !gppSection ||
+    (gppRegion === US_NATIONAL_REGION && usApproach === GPPUSApproach.STATE)
+  ) {
+    // If we don't have a section, we can't get anything.
+    // If we're using the state approach, we shouldn't use the national section, even if region is set to national.
+    return consent;
+  }
+
+  notices.forEach((notice) => {
+    const { notice_key: noticeKey, gpp_field_mapping: fieldMapping } = notice;
+    const gppMechanisms = fieldMapping?.find(
+      (fm) => fm.region === gppRegion,
+    )?.mechanism;
+
+    if (gppMechanisms) {
+      // For each mechanism, check the field value and determine the consent value
+      const hasOptOut = gppMechanisms.some((gppMechanism) => {
+        const fieldValue = cmpApi.getFieldValue(
+          gppSection.name,
+          gppMechanism.field,
+        );
+
+        // Convert field value to string for comparison
+        const valueStr = Array.isArray(fieldValue)
+          ? fieldValue.join("")
+          : String(fieldValue);
+
+        // Check if the value matches any of the defined states
+        if (valueStr === gppMechanism.opt_out) {
+          return true; // User has opted out
+        }
+        if (valueStr === gppMechanism.not_opt_out) {
+          return false; // User has not opted out
+        }
+        // If value matches not_available or is undefined, we'll skip this mechanism
+        return undefined;
```

Comment:
We're in a `some` fn which only looks at true / false, otherwise converts `undefined` to `false`- Just making sure we want to use the implicit `undefined` here?

Reply:
Good catch!